src/bits/find_if.h

```diff
@@ -84,17 +84,63 @@ class iterator_result
     Iterator end;
 };
 
+template <typename Iterator>
+class iterator_result_rvalue
+{
+public:
+    iterator_result_rvalue(Iterator iterator, Iterator end)
+        : m_has_value(iterator != end)
+    {
+        if (m_has_value)
+            m_value = std::move(*iterator);
+    }
+    operator bool() const { return m_has_value; }
+    bool has_result() const { return m_has_value; }
+    decltype(auto) operator*() const
+    {
+        assert(m_has_value);
+        return m_value;
+    }
+    decltype(auto) operator->() const
+    {
+        assert(m_has_value);
+        return &m_value;
+    }
+
+    bool m_has_value;
+    typename Iterator::value_type m_value;
```

Comment:
Because this isn't initialized in the initializer-list, it requires that Iterator::value_type must be default-constructeable.
Also, with m_has_value, this is basically just re-creating std::optional.
Ideally just use `std::optional<Iterator::value_type>` here directly.

Reply:
std::optional is C++17, and KDAlgorithms should work with C++14.
Indeed it requires the value to be default construcable. I guess I could get around that by storing it using a pointer, wonder if it is worthwhile going that far.

Reply:
Ah, okay, was expecting it to require C++17 already.
Hm, you could just use a unique_ptr I guess :thinking: 
Would require an additional heap allocation, but could then be very easily replaced by std::optional when upgrading to c++17